torch/csrc/autograd/custom_function.h

```diff
@@ -11,4 +11,197 @@ TORCH_API variable_list _wrap_outputs(
+enable_if_var_t<T> extract_vars(std::vector<bool> &is_var, variable_list& list, T&& cur, Args&& ... args) {
+  is_var.push_back(true);
+  list.emplace_back(cur);
+  extract_vars(is_var, list, std::forward<Args>(args)...);
+}
+
+template <typename... Args>
+void extract_vars(std::vector<bool> &is_var, variable_list& list, Args&& ... args) {
+}
+
+template<class T>
+template<typename... Args>
+variable_list Function<T>::apply(Args&&... args) {
+  std::shared_ptr<CppNode<T>> node(new CppNode<T>, deleteNode);
+  variable_list input_vars;
+
+  const size_t num_inputs = sizeof...(Args);
+  input_vars.reserve(num_inputs);
+  node->is_variable_input.reserve(num_inputs);
+
+  extract_vars(node->is_variable_input, input_vars, args...);
+
+  bool is_executable =  GradMode::is_enabled() && any_variable_requires_grad(input_vars);
+  auto next_edges = collect_next_edges(input_vars);
+  set_ctx_grad_fn(node->ctx, node);
+  node->set_next_edges(std::move(next_edges));
+  node->clear_input_metadata();
+
+  node->input_info.reserve(input_vars.size());
+  for (auto& var : input_vars) {
+      node->input_info.emplace_back(var);
+  }
+
+  variable_list outputs;
+  {
+    AutoGradMode grad_mode(false);
+    outputs = T::forward(&node->ctx, std::forward<Args>(args)...);
+  }
+
+  auto wrapped_outputs = _wrap_outputs(input_vars, node->ctx.get_non_differentiable(), node->ctx.get_dirty(), outputs, is_executable ? node : nullptr);
+
+  node->output_info.reserve(wrapped_outputs.size());
+  for (auto& output : wrapped_outputs) {
+    if (is_executable) {
+      node->output_info.emplace_back(output);
+    }
+  }
+
+  return wrapped_outputs;
+}
+
+template<class T>
+variable_list CppNode<T>::apply(variable_list&& inputs) {
+  at::OptionalDeviceGuard _device_guard;
+
+  int num_inputs = inputs.size();
+  variable_list backward_inputs;
+  backward_inputs.reserve(num_inputs);
+  for (int i = 0 ; i < num_inputs; ++i) {
+    if (inputs[i].defined()) {
+      backward_inputs.emplace_back(inputs[i]);
+    } else {
+      backward_inputs.emplace_back(output_info[i].zeros(_device_guard));
+    }
+  }
+
+  auto outputs = T::backward(&ctx, backward_inputs);
+
+  int num_forward_inputs = is_variable_input.size();
+  int num_outputs = outputs.size();
+  // Returning too many results is ok, but only as long as they're all undefined.
+  // Truncate the result vector in that case.
+  if (num_outputs > num_forward_inputs) {
+    bool all_undef = true;
+    for (int i = num_forward_inputs; i < num_outputs; ++i) {
+      all_undef &= (!outputs[i].defined());
+    }
+    if (all_undef) {
+      outputs.resize(num_forward_inputs);
+      num_outputs = num_forward_inputs;
+    }
+  }
+
+  if (num_outputs != num_forward_inputs) {
+    std::string msg("function ");
+    msg += name() + " returned an incorrect number of gradients (expected ";
+    msg += std::to_string(num_forward_inputs) + ", got " ;
+    msg += std::to_string(num_outputs) + ")";
+    throw std::runtime_error(msg);
```

Comment:
I know you lifted this straight from the old code (so, no action needed in this PR), but we can modernize the error reporting here, using the `TORCH_CHECK` macro.